File: ios/Vxin/Core/Storage/DraftStore.swift
import Foundation

/// Drafts are isolated by server, account and conversation. Legacy unowned drafts are not read.
final class DraftStore {
    static let shared = DraftStore()
    private let defaults: UserDefaults
    private let currentScope: () -> MessageScope?
    private let prefix = "vxin_draft_"

    init(defaults: UserDefaults = .standard, currentScope: @escaping () -> MessageScope? = { MessageScope.current }) {
        self.defaults = defaults
        self.currentScope = currentScope
    }

    func get(_ conversationId: String, scope: MessageScope? = MessageScope.current) -> String {
        guard let scope, scope == currentScope(), !conversationId.isEmpty else { return "" }
        return defaults.string(forKey: prefix + scope.key + ":" + conversationId) ?? ""
    }

    func set(_ conversationId: String, _ text: String, scope: MessageScope? = MessageScope.current) {
        guard let scope, scope == currentScope(), !conversationId.isEmpty else { return }
        let key = prefix + scope.key + ":" + conversationId
        if text.isEmpty { defaults.removeObject(forKey: key) }
        else { defaults.set(text, forKey: key) }
    }

    func clear(_ conversationId: String, scope: MessageScope? = MessageScope.current) { set(conversationId, "", scope: scope) }

    func clearScope(_ scope: MessageScope? = MessageScope.current) {
        for key in defaults.dictionaryRepresentation().keys where key.hasPrefix(prefix) {
            if !key.hasPrefix(prefix + "v2_") || scope.map({ key.hasPrefix(prefix + $0.key + ":") }) == true {
                defaults.removeObject(forKey: key)
            }
        }
    }
}

File: android/app/src/main/java/com/vxin/app/core/storage/DraftStore.kt
package com.vxin.app.core.storage

import android.content.Context
import android.content.SharedPreferences
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton

/** Drafts are isolated by server, account and conversation; old unowned keys are never read. */
@Singleton
class DraftStore internal constructor(
    private val prefs: SharedPreferences,
    private val scopes: MessageScopeProvider,
) {
    @Inject constructor(@ApplicationContext context: Context, scopes: MessageScopeProvider) :
        this(context.getSharedPreferences("vxin_drafts", Context.MODE_PRIVATE), scopes)

    fun get(conversationId: String, scope: MessageScope? = scopes.current()): String {
        if (conversationId.isBlank() || !scopes.isCurrent(scope)) return ""
        return runCatching { prefs.getString("${scope!!.key}:$conversationId", "").orEmpty() }.getOrDefault("")
    }

    fun set(conversationId: String, text: String, scope: MessageScope? = scopes.current()) {
        if (conversationId.isBlank() || !scopes.isCurrent(scope)) return
        val key = "${scope!!.key}:$conversationId"
        prefs.edit().apply { if (text.isEmpty()) remove(key) else putString(key, text) }.apply()
    }

    fun clear(conversationId: String, scope: MessageScope? = scopes.current()) = set(conversationId, "", scope)

    fun clearScope(scope: MessageScope? = scopes.current()) {
        val prefix = scope?.let { "${it.key}:" }
        prefs.edit().apply {
            prefs.all.keys.filter { !it.startsWith("v2_") || (prefix != null && it.startsWith(prefix)) }
                .forEach { remove(it) }
        }.apply()
    }
}
